scylla/src/transport/node.rs

```diff
@@ -55,6 +57,72 @@ impl TimestampedAverage {
     }
 }
 
+/// This enum is introduced to support address translation only in PoolRefiller,
+/// as well as to cope with the bug in older Cassandra and Scylla releases.
+/// The bug involves misconfiguration of rpc_address and/or broadcast_rpc_address
+/// in system.local to 0.0.0.0. Mitigation involves replacing the faulty address
+/// with connection's address, but then that address must not be subject to AddressTranslator,
+/// so we carry that information using this enum. Address translation is never performed
+/// on InitiallyKnown and FixedLocal variants.
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy)]
+pub enum NodeAddr {
+    RealPeer(SocketAddr),       // fetched in Metadata with query_peers()
+    InitiallyKnown(SocketAddr), // provided by user in SessionConfig
+    FixedLocal(SocketAddr), // built from control connection's address in order to mitigate the bug described above
+}
+impl NodeAddr {
+    pub(crate) fn inner(&self) -> SocketAddr {
+        match self {
+            NodeAddr::RealPeer(addr)
+            | NodeAddr::InitiallyKnown(addr)
+            | NodeAddr::FixedLocal(addr) => *addr,
+        }
+    }
+    pub fn ip(&self) -> IpAddr {
+        self.inner().ip()
+    }
+    pub fn port(&self) -> u16 {
+        self.inner().port()
+    }
+}
+impl Display for NodeAddr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.inner())
+    }
+}
+impl From<SocketAddr> for NodeAddr {
+    fn from(addr: SocketAddr) -> Self {
+        Self::RealPeer(addr)
+    }
+}
+impl From<&SocketAddr> for NodeAddr {
+    fn from(addr: &SocketAddr) -> Self {
+        Self::RealPeer(*addr)
+    }
+}
+impl PartialEq for NodeAddr {
+    fn eq(&self, other: &Self) -> bool {
+        self.inner() == other.inner()
+    }
+}
+impl Eq for NodeAddr {}
+impl PartialOrd for NodeAddr {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.inner().partial_cmp(&other.inner())
+    }
+}
+impl Ord for NodeAddr {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.inner().cmp(&other.inner())
+    }
+}
+impl Hash for NodeAddr {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.inner().hash(state);
+    }
+}
```

Comment:
I can't recall what made me choose this way. One idea: in case of no AddressTranslator provided, with current `known_peers` implementation, `Node`s will be recycled instead of being rebuilt if a control connection changes to another node: both previous and new control connection's `Node`s will be spared.